packages/amplify-graphql-model-transformer/src/resources/custom-resource-lambda/custom-resource-handler.ts

```diff
@@ -0,0 +1,440 @@
+  IsComplete: true,
+};
+const notFinished: AWSCDKAsyncCustomResource.IsCompleteResponse = {
+  IsComplete: false,
+};
+
+// compares the currentState with the endState to determine a next update step that will get the table closer to the end state
+export const getNextGSIUpdate = (currentState: TableDescription, endState: CustomDDB.Input): UpdateTableInput | undefined => {
+  const endStateGSIs = endState.GlobalSecondaryIndexes || [];
+  const endStateGSINames = endStateGSIs.map((gsi) => gsi.IndexName);
+
+  const currentStateGSIs = currentState.GlobalSecondaryIndexes || [];
+  const currentStateGSINames = currentStateGSIs.map((gsi) => gsi.IndexName);
+
+  // function to identify any GSIs that need to be removed
+  const gsiRequiresReplacementPredicate = (currentGSI: DynamoDB.GlobalSecondaryIndexDescription): boolean => {
+    // check if the index has been removed entirely
+    if (!endStateGSINames.includes(currentGSI.IndexName!)) return true;
+    // get the end state of this GSI
+    const respectiveEndStateGSI = endStateGSIs.find((endStateGSI) => endStateGSI.IndexName === currentGSI.IndexName)!;
+    // detect if projection has changed
+    if (isProjectionModified(currentGSI.Projection!, respectiveEndStateGSI.Projection!)) return true;
+    // detect if key schema has changed
+    if (isKeySchemaModified(currentGSI.KeySchema!, respectiveEndStateGSI.KeySchema!)) return true;
+    // if we got here, then the GSI does not need to be removed
+    return false;
+  };
+  const gsiToRemove = currentStateGSIs.find(gsiRequiresReplacementPredicate);
+  if (gsiToRemove) {
+    return {
+      TableName: currentState.TableName!,
+      GlobalSecondaryIndexUpdates: [
+        {
+          Delete: {
+            IndexName: gsiToRemove.IndexName!,
+          },
+        },
+      ],
+    };
+  }
+
+  // if we get here, then find a GSI that needs to be created and construct an update request
+  const gsiRequiresCreationPredicate = (endStateGSI: DynamoDB.GlobalSecondaryIndex): boolean =>
+    !currentStateGSINames.includes(endStateGSI.IndexName);
+
+  const gsiToAdd = endStateGSIs.find(gsiRequiresCreationPredicate);
+  if (gsiToAdd) {
+    const attributeNamesToInclude = gsiToAdd.KeySchema.map((schema) => schema.AttributeName);
+    const gsiToAddAction = removeUndefinedAttributes({
+      IndexName: gsiToAdd.IndexName,
+      KeySchema: gsiToAdd.KeySchema,
+      Projection: gsiToAdd.Projection,
+      ProvisionedThroughput: gsiToAdd.ProvisionedThroughput,
+    }) as DynamoDB.CreateGlobalSecondaryIndexAction;
+    return {
+      TableName: currentState.TableName!,
+      AttributeDefinitions: endState.AttributeDefinitions.filter((def) => attributeNamesToInclude.includes(def.AttributeName)),
+      GlobalSecondaryIndexUpdates: [
+        {
+          Create: gsiToAddAction,
+        },
+      ],
+    };
+  }
+
+  // no more updates necessary
+  return undefined;
+};
+
+type CreateTableResponse = {
```

Comment:
Is this factorable into a common type def, since the CDK code is going to operate on the type as well?

Reply:
This type is used by lambda only. What do you mean by the CDK code usage?

Reply:
Doesn't the CDK code that used the custom resource to create the table need the CreateTableResponse (e.g., the table name) in order to complete its work setting up resolvers, etc? If so, is that information delivered in some other structure than the CreateTableResponse?

Reply:
There are two steps happening. First is that the create table response is converted to the CFN response in the lambda which is as following
```
      const response = await createNewTable(createTableInput);
      result = {
        PhysicalResourceId: response.tableName,
        Data: {
          TableArn: response.tableArn,
          TableStreamArn: response.streamArn,
          TableName: response.tableName,
        },
      };
```
The next step is that a `Table.fromTableAttributes` is called to construct a table wrapper so that it can be passed into the resolver generation functions. The way to extract the attributes from CFN response is by using intrinsic function `Fn::GetAttr` (https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/crpg-ref-responses.html)
```
    const table = Table.fromTableAttributes(scope, `AmplifyTable${tableLogicalName}`, {
      tableArn: tableResource.getAttString('TableArn'),
      tableStreamArn: tableResource.getAttString('TableStreamArn'),
    });
```

Reply:
The related type here is the `OnEventResponse`. Currently it is more of a generic type. I can add more specific type definitions for the tables in the upcoming PRs.